File: app/src/main/java/com/platdmit/simplecloudmanager/domain/helpers/ContentUpdateService.java
package com.platdmit.simplecloudmanager.domain.helpers;

import com.platdmit.simplecloudmanager.domain.repo.UpdateScheduleRepo;

import org.joda.time.DateTime;

import java.util.HashMap;
import java.util.Map;

public class ContentUpdateService implements UpdateScheduleService{
    private static final String TAG = ContentUpdateService.class.getSimpleName();
    private int mLifeTime = 900000;
    private UpdateScheduleRepo mUpdateScheduleRepo;
    private Map<String, DateTime> mActiveTypes;

    public ContentUpdateService(UpdateScheduleRepo updateScheduleRepo) {
        mUpdateScheduleRepo = updateScheduleRepo;
        mActiveTypes = new HashMap<>();
    }

    public ContentUpdateService(UpdateScheduleRepo updateScheduleRepo, int defaultLifeTime) {
        mUpdateScheduleRepo = updateScheduleRepo;
        mActiveTypes = new HashMap<>();
        mLifeTime = defaultLifeTime;
    }

    @Override
    public void setUpdateTime(String key, int timer) {
        DateTime updateTime = new DateTime().plusMillis(timer);
        mActiveTypes.put(key, updateTime);
        mUpdateScheduleRepo.setUpdateTime(key, updateTime.getMillis());
    }

    @Override
    public void setDefaultUpdateTime(String key) {
        setUpdateTime(key, mLifeTime);
    }

    @Override
    public boolean getActualStatus(String key) {
        long updateTime = getUpdateTime(key);
        return new DateTime().getMillis() < updateTime;
    }

    public long getUpdateTime(String key) {
        DateTime updateTime = mActiveTypes.get(key);
        if(updateTime != null){
            return updateTime.getMillis();
        } else {
            long dbUpdateTime = mUpdateScheduleRepo.getUpdateTime(key);
            if(dbUpdateTime > 0){
                mActiveTypes.put(key, new DateTime(dbUpdateTime));
            }
            return dbUpdateTime;
        }
    }
}

File: domain/src/main/java/com/platdmit/domain/helpers/ContentUpdateService.kt
package com.platdmit.domain.helpers

import com.platdmit.domain.repo.UpdateScheduleRepo
import org.joda.time.DateTime
import java.util.*

class ContentUpdateService(
        private val updateScheduleRepo: UpdateScheduleRepo,
        private val lifeTime: Int = 900000
) : UpdateScheduleService  {
    private val activeTypes: MutableMap<String, DateTime> = HashMap()

    override fun setUpdateTime(key: String, timer: Int) {
        val updateTime = DateTime().plusMillis(timer)
        activeTypes[key] = updateTime
        updateScheduleRepo.setUpdateTime(key, updateTime.millis)
    }

    override fun setDefaultUpdateTime(key: String) {
        setUpdateTime(key, lifeTime)
    }

    override fun getActualStatus(key: String): Boolean {
        val updateTime = getUpdateTime(key)
        return DateTime().millis < updateTime
    }

    private fun getUpdateTime(key: String): Long {
        val updateTime = activeTypes[key]
        return if (updateTime != null) {
            updateTime.millis
        } else {
            val dbUpdateTime = updateScheduleRepo.getUpdateTime(key)
            if (dbUpdateTime > 0) {
                activeTypes[key] = DateTime(dbUpdateTime)
            }
            dbUpdateTime
        }
    }

    companion object {
        private val TAG = ContentUpdateService::class.java.simpleName
    }
}